functorch/_src/aot_autograd.py

```diff
@@ -506,24 +555,119 @@ def aot_module(mod: nn.Module, *args, **kwargs) -> nn.Module:
+        # I'm not actually sure if this is the right thing to do ...
+        def create_arg(self, a):
+            if isinstance(a, torch.nn.Parameter):
+                for n, p in self.root.named_parameters():
+                    if a is p:
+                        return self.create_node('get_attr', n, (), {})
+                qualname = None
+
+                if not qualname:
+                    i = 0
+                    while True:
+                        qualname = f'_param_constant{i}'
+                        if not hasattr(self.root, qualname):
+                            break
+                        i += 1
+                    setattr(self.root, qualname, a)
+
+                return self.create_node('get_attr', qualname, (), {})
+            return super().create_arg(a)
+    
+    def flattened_fn(gm, nargs):
+        # Lets flatten the params
+        # Params value will change over time but spec will remain same
+
+        params = {
+            **dict(_named_parameters(gm, remove_duplicate=False)),
+            **dict(_named_buffers(gm, remove_duplicate=False)),
+        }
+        params_flat, params_spec = pytree.tree_flatten(params)
+        params_flat = tuple(params_flat)
+        params_len = len(params_flat)
+
+        def fn_for_tracing(*proxy_args):
+            assert len(proxy_args) == nargs
+            with _stateless.reparametrize_module(
+                gm, pytree.tree_unflatten(proxy_args[:params_len], params_spec)
+            ):
+                # out = PatchingInterpreter(gm).run(*args[params_len:])
+                out = gm(*proxy_args[params_len:])
+            assert isinstance(out, (tuple, list)), "graph must output tuple()"
+            return out
+
+        tracer: torch.fx.Tracer = MyTracer()
+        graph = tracer.trace(fake_signature(fn_for_tracing, nargs))
+        traced = torch.fx.GraphModule(tracer.root, graph, "my_tracer")
+        traced.recompile()
+
+
+        def fn_with_params_as_args(*joint_args):
+            return traced(*joint_args)
+        return fn_with_params_as_args
+
+
+    gm = torch.fx.symbolic_trace(mod)
+    compiled_f = None
+
 
     class AOTModule(nn.Module):
         def __init__(self):
             super(AOTModule, self).__init__()
             self.orig_module = mod
 
         def forward(self, *args, **kwargs):
-            return compiled_f(
-                dict(_named_parameters(mod, remove_duplicate=False)),
-                dict(_named_buffers(mod, remove_duplicate=False)),
-                *args,
-                **kwargs,
-            )
+            nonlocal compiled_f
+
+            params = {
+                **dict(_named_parameters(gm, remove_duplicate=False)),
+                **dict(_named_buffers(gm, remove_duplicate=False)),
+            }
+            if HAS_TREE:
+                params_flat = tree.flatten(params)
+            else:
+                params_flat, _ = pytree.tree_flatten(params)
+
+            params_flat = tuple(params_flat)
```

Comment:
In the torchdynamo case (not in the general case), this list is a constant you don't need to recompute every time.

Reply:
For training, the params will change after each update. Don't we have to flatten for each call?

Reply:
Yeah, I realized while driving today. These are just references. And the update happens in place.

Will remove it.